### PR TITLE
CLI: support for adding more subkeys. Closes #1784.

### LINKED ISSUE #1784: CLI: rnpkeys does not support adding new additional subkeys to existing primary keys.
## Description

Looks like this basic & important feature was discussed and planned to be implemented long time ago (like here https://github.com/rnpgp/rnp/issues/323) but is not implemented yet.

## Steps to Reproduce

There is no `rnpkeys` command to add new subkeys. One has to use GnuPG command line tool for this task.

## Expected Behavior

`rnpkeys` should support adding subkeys. This feature can be implemented in multiple ways:

- Implement new command `--add-subkey` with primary key specified as a parameter.
- Add new option `--subkey` to the `--generate-key` command.
- Extend `--edit-key` command functionality.

### PR DESCRIPTION
**Pull request by antonsviridenko**
Implements new feature in `rnpkeys`, now more subkeys can be generated and added to existing primary keys using `--edit-key` `--add-subkey` options.

### CODE REVIEW

File: src/rnp/fficli.cpp [hunk 1]
```diff
@@ -958,6 +958,94 @@ cli_rnp_t::fix_cv25519_subkey(const std::string &key, bool checkonly)
...
+    // TODO : set DSA qbits
+    if (rnp_op_generate_set_hash(genkey, cfg().get_cstr(CFG_KG_HASH))) {
+        ERR_MSG("Failed to set hash algorithm.");
+        goto done;
+    }
+    if (rnp_op_generate_execute(genkey) || rnp_op_generate_get_key(genkey, &subkey)) {
+        ERR_MSG("Subkey generation failed.");
+        goto done;
+    }
+    if (rnp_request_password(ffi, subkey, "protect", &password)) {
```
Comment (ni4) [inline src/rnp/fficli.cpp:1015]: As most likely user would need to enter corresponding secret's key password, we should ask whether he wants to use the same password for the subkey. However this could be extracted to the separate task.

File: src/rnpkeys/rnpkeys.cpp [hunk 2]
```diff
@@ -68,6 +68,12 @@ const char *usage =
   "  --revoke-key           Revoke a key specified.\n"
   "  --remove-key           Remove a key specified.\n"
   "  --edit-key             Edit key properties.\n"
+  "    --add-subkey         Add new subkey.\n"
+  "        --expert         Select subkey type, size and additional parameters.\n"
```
Comment (ni4) [inline src/rnpkeys/rnpkeys.cpp:72]: Sub-options would duplicate the --generate-key options, making this list too long. Probably we should remove them, allowing user to read man in case he needs it.

File: src/rnpkeys/tui.cpp [hunk 3]
```diff
@@ -152,6 +152,28 @@ ask_rsa_bitlen(FILE *input_fp)
     return result;
 }
 
+static long
+ask_elgamal_bitlen(FILE *input_fp)
+{
+    do {
+        printf(
```
Comment (ni4) [inline src/rnpkeys/tui.cpp:159]: It would be good to cover these lines via cli_tests.

File: src/rnpkeys/tui.cpp [hunk 4]
```diff
@@ -152,6 +152,28 @@ ask_rsa_bitlen(FILE *input_fp)
...
+          ELGAMAL_MIN_P_BITLEN,
+          ELGAMAL_MAX_P_BITLEN,
+          DEFAULT_ELGAMAL_NUMBITS);
+        long result = DEFAULT_ELGAMAL_NUMBITS;
+        if (!rnp_secure_get_long_from_fd(input_fp, result)) {
+            continue;
+        }
+        if ((result >= ELGAMAL_MIN_P_BITLEN) && (result <= ELGAMAL_MAX_P_BITLEN)) {
+            // round up to multiple of 32
+            result = ((result + 31) / 32) * 32;
```
Comment (ni4) [inline src/rnpkeys/tui.cpp:170]: More common in cryptography-related software would be to see `result = ((result + 0x1f) >> 5) << 5` :)

**Review by ni4 — COMMENTED** [review]
Please see the review comments.

File: src/rnpkeys/tui.cpp (hunk 4 above)
Comment (antonsviridenko) [reply src/rnpkeys/tui.cpp:170]: I used existing similar code from `ask_dsa_bitlen(FILE *input_fp)` as a reference:
https://github.com/rnpgp/rnp/blob/master/src/rnpkeys/tui.cpp#L170

File: src/rnp/fficli.cpp (hunk 1 above)
Comment (antonsviridenko) [reply src/rnp/fficli.cpp:1018]: done

**Comment by antonsviridenko** [comment]
No SM2 with openssl :( probably should not be listed in available key algorithms then...

File: src/rnpkeys/tui.cpp (hunk 4 above)
Comment (ni4) [reply src/rnpkeys/tui.cpp:170]: Ok, let it be as it is then, probably some minor refactoring PR could fix these afterwards.

**Comment by ni4** [comment]
> No SM2 with openssl :( probably should not be listed in available key algorithms then...

Yeah, we should check for it's presence via `rnp_supports_feature()`. However, if it is not checked for the primary key generation as well this can be moved to a separate PR.

**Comment by antonsviridenko** [comment]
@ni4 ok, do you approve it, is it ok to merge?

**Review by ronaldtse — APPROVED** [review]

**Comment by antonsviridenko** [comment]
only codecov test failed

File: src/rnp/fficli.cpp [hunk 5]
```diff
@@ -463,7 +466,7 @@ ffi_pass_callback_stdin(rnp_ffi_t        ffi,
             }
         }
     }
-    if (protect && is_primary) {
+    if ((protect | add_subkey) && is_primary) {
```
Comment (ni4) [inline src/rnp/fficli.cpp:469]: Should not we use logical or here (i.e. `||`)?

File: src/rnp/fficli.cpp [hunk 6]
```diff
@@ -394,6 +395,8 @@ ffi_pass_callback_stdin(rnp_ffi_t        ffi,
 
     if (!strcmp(pgp_context, "protect")) {
         protect = true;
+    } else if (!strcmp(pgp_context, "add subkey")) {
+        add_subkey = true;
```
Comment (ni4) [inline src/rnp/fficli.cpp:399]: This and following lines for `add_subkey` case should be covered with tests, too major functionality to skip.

**Review by ni4 — COMMENTED** [review]
Please see the review comments. It would be good to have some more tests coverage where possible. It could be wrong bits size, wrong curve type/algorithm mismatch, whatever else.

File: src/rnp/fficli.cpp (hunk 5 above)
Comment (antonsviridenko) [reply src/rnp/fficli.cpp]: fixed

File: src/rnp/fficli.cpp (hunk 6 above)
Comment (antonsviridenko) [reply src/rnp/fficli.cpp:399]: see `test_additional_subkeys_reuse_password` in cli_tests.py

**Comment by antonsviridenko** [comment]
@ni4 
> It could be wrong bits size, wrong curve type/algorithm mismatch, whatever else.
It is alredy tested in
```
def test_additional_subkeys_expert_mode(self):
        # Run primary key generation
        ret, _, _ = run_proc(RNPK, ['--homedir', RNPDIR, '--password=',
                                        '--userid', 'primary_for_many_subs@rnp', '--generate-key'])
        # RSA subkey
        ret, out, err = run_proc(RNPK, ['--homedir', RNPDIR, '--password=', '--edit-key', '--add-subkey', 'primary_for_many_subs@rnp',
                                        '--expert'], '\n\n0\n101\n1\n1023\n4097\n3072\n')
        self.assertEqual(ret, 0, KEY_GEN_FAILED)
        # ElGamal subkey
        ret, out, err = run_proc(RNPK, ['--homedir', RNPDIR, '--password=', '--edit-key', '--add-subkey', 'primary_for_many_subs@rnp',
                                        '--expert'], '\n\n0\n101\n16\n1023\n4097\n1025\n')
        self.assertEqual(ret, 0, KEY_GEN_FAILED)
        # DSA subkey
        ret, out, err = run_proc(RNPK, ['--homedir', RNPDIR, '--password=', '--edit-key', '--add-subkey', 'primary_for_many_subs@rnp',
                                        '--expert'], '\n\n0\n101\n17\n1023\n3073\n1025\n')
        self.assertEqual(ret, 0, KEY_GEN_FAILED)
        # ECDH subkey
        ret, out, err = run_proc(RNPK, ['--homedir', RNPDIR, '--password=', '--edit-key', '--add-subkey', 'primary_for_many_subs@rnp',
                                        '--expert'], '\n\n0\n101\n18\n0\n8\n1\n')
        self.assertEqual(ret, 0, KEY_GEN_FAILED)
        # ECDSA subkey
        ret, out, err = run_proc(RNPK, ['--homedir', RNPDIR, '--password=', '--edit-key', '--add-subkey', 'primary_for_many_subs@rnp',
                                        '--expert'], '\n\n0\n101\n19\n0\n8\n1\n')
        self.assertEqual(ret, 0, KEY_GEN_FAILED)
        # EDDSA subkey
        ret, out, err = run_proc(RNPK, ['--homedir', RNPDIR, '--password=', '--edit-key', '--add-subkey', 'primary_for_many_subs@rnp',
                                        '--expert'], '\n\n0\n101\n22\n')
```
see stdin string passed to the process, wrong algorithm number is tried first, then correct one, then invalid bit length or curve number, then correct one

**Comment by antonsviridenko** [comment]
ok, rebased against the master, now shellcheck test is successfull, only codecov fails

File: src/rnp/fficli.cpp (hunk 6 above)
Comment (ni4) [reply src/rnp/fficli.cpp:399]: Thanks!

**Comment by ni4** [comment]
@antonsviridenko I pushed commit with some more edge cases checks, hope it will make codecov happier.

**Comment by antonsviridenko** [comment]
Does cli_test affect codecov results? I though only compiled tests do...

**Comment by ni4** [comment]
> Does cli_test affect codecov results? I though only compiled tests do...

Yeah, I also thought so, like it investigates code paths during compilations or suchlike stuff. But in reality it's simpler: all code, compiled with `--coverage`, logs execution code lines info to the `*.gcov` files, and then those are uploaded to the codecov.io (see `success.sh` for ghis step). The only thing I don't know for sure now is whether we may upload multiple *.gcov sets there (i.e. for Windows runs, OpenSSL runs, and so on).

**Comment by antonsviridenko** [comment]
@ni4 all green :)

**Review by ni4 — APPROVED** [review]
LGTM, thanks!
@antonsviridenko  Feel free to merge if you agree with my additions.
